base/random/RNGs.jl

```diff
@@ -137,7 +137,7 @@ hash(r::MersenneTwister, h::UInt) = foldr(hash, h, (r.seed, r.state, r.vals, r.i
 @inline mt_pop!(r::MersenneTwister) = @inbounds return r.vals[r.idx+=1]
 
 function gen_rand(r::MersenneTwister)
-    dsfmt_fill_array_close1_open2!(r.state, pointer(r.vals), length(r.vals))
+    Base.@gc_preserve r dsfmt_fill_array_close1_open2!(r.state, pointer(r.vals), length(r.vals))
```

Comment:
Here (also in other places), the "preserved" object is again used on the line after, isn't it enough to preserve it? (I guess not as you updated the code) Is it that the compiler can reorder instructions?

Reply:
No it's not.

Reply:
To clarify, after inlining any uses after may be dead code and eliminated.

Reply:
This is the detail I don't want to go into since I don't want people to rely on it but another way this can be invalid is that the compiler can see you are using only some fields and then discards all other fields.